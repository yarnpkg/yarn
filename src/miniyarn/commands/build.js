import {flags} from '@manaflair/concierge';

export default concierge =>
  concierge
    .command(`build [... args]`)
    .describe(`Run the project build script, if available`)
    .flag(flags.PROXY_COMMAND)
    .action(async args => {
      return await concierge.run(null, [`run`, `build`, ...args.args], args);
    });
