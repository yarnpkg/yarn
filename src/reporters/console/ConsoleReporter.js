/**
 * Copyright (c) 2016-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @flow
 */

import BaseReporter from '../BaseReporter.js';
import Progress from './ProgressBar.js';
import Spinner from './SpinnerProgress.js';
import type {Package, Trees, ReporterSpinner} from '../types.js';
import {clearLine} from './util.js';
import {removeSuffix} from '../../util/misc.js';

let readline = require('readline');
let repeat = require('repeating');
let chalk = require('chalk');
let read = require('read');
let _ = require('lodash');

function sortTrees(trees: Trees = []): Trees {
  return trees.sort(function(tree1, tree2): number {
    return tree1.name.localeCompare(tree2.name);
  });
}

export default class ConsoleReporter extends BaseReporter {
  _prependEmoji(msg: string, emoji: ?string): string {
    if (this.emoji && emoji && this.isTTY) {
      msg = `${emoji}  ${msg}`;
    }
    return msg;
  }

  step(current: number, total: number, msg: string, emoji?: string) {
    msg = this._prependEmoji(msg, emoji);

    if (_.endsWith(msg, '?')) {
      msg = `${removeSuffix(msg, '?')}...?`;
    } else {
      msg += '...';
    }

    this.log(`${chalk.grey(`[${current}/${total}]`)} ${msg}`);
  }

  list(key: string, items: Array<string>) {
    for (let item of items) {
      this.log(`   - ${item}`);
    }
  }

  header(command: string, pkg: Package) {
    this.log(chalk.bold(`${pkg.name} ${command} v${pkg.version}`));
  }

  footer(showPeakMemory?: boolean) {
    let totalTime = (this.getTotalTime() / 1000).toFixed(2);
    let msg = `Done in ${totalTime}s.`;
    if (showPeakMemory) {
      let peakMemory = (this.peakMemory / 1024 / 1024).toFixed(2);
      msg += ` Peak memory usage ${peakMemory}MB.`;
    }
    this.log(this._prependEmoji(msg, '✨'));
  }

  log(msg: string) {
    clearLine(this.stdout);
    this.stdout.write(`${msg}\n`);
  }

  success(msg: string) {
    this.log(`${chalk.green('success')} ${msg}`);
  }

  error(msg: string) {
    clearLine(this.stderr);
    this.stderr.write(`${chalk.red('error')} ${msg}\n`);
  }

  info(msg: string) {
    this.log(`${chalk.blue('info')} ${msg}`);
  }

  command(command: string) {
    this.log(chalk.grey(`$ ${command}`));
  }

  warn(msg: string) {
    clearLine(this.stderr);
    this.stderr.write(`${chalk.yellow('warning')} ${msg}\n`);
  }

  question(question: string, password?: boolean): Promise<string> {
    if (!process.stdout.isTTY) {
      return Promise.reject(new Error("Can't answer a question unless a user TTY"));
    }

    return new Promise((resolve, reject) => {
      read({
        prompt: `${chalk.grey('question')} ${question}: `,
        silent: !!password,
        output: this.stdout,
        input: this.stdin,
      }, (err, answer) => {
        if (err) {
          reject(err);
        } else {
          resolve(answer);
        }
      });
    });
  }

  tree(key: string, trees: Trees) {
    trees = sortTrees(trees);

    let stdout = this.stdout;

    function output({name, children, hint, color}, level, end) {
      children = sortTrees(children);

      let indent = end ? '└' : '├';

      if (level) {
        indent = repeat('│  ', level) + indent;
      }

      let suffix = '';
      if (hint) {
        suffix += ` (${chalk.grey(hint)})`;
      }
      if (color) {
        name = chalk[color](name);
      }
      stdout.write(`${indent}─ ${name}${suffix}\n`);

      if (children && children.length) {
        for (let i = 0; i < children.length; i++) {
          let tree = children[i];
          output(tree, level + 1, i === children.length - 1);
        }
      }
    }

    for (let i = 0; i < trees.length; i++) {
      let tree = trees[i];
      output(tree, 0, i === trees.length - 1);
    }
  }

  activityStep(current: number, total: number, msg: string, emoji?: string): ReporterSpinner {
    if (!this.isTTY) {
      return this.activity();
    }

    msg = this._prependEmoji(msg, emoji);

    let spinner = new Spinner(this.stderr, `${chalk.grey(`[${current}/${total}]`)} `);
    spinner.start();
    spinner.setText(msg);

    return {
      tick(name: string) {
        spinner.setText(`${msg}: ${name}`);
      },

      end() {
        spinner.stop();
      },
    };
  }

  activity(): ReporterSpinner {
    if (!this.isTTY) {
      return {
        tick() {},
        end() {},
      };
    }

    let spinner = new Spinner(this.stderr);
    spinner.start();

    return {
      tick(name: string) {
        spinner.setText(name);
      },

      end() {
        spinner.stop();
      },
    };
  }

  select(header: string, question: string, options: Array<string>): Promise<string> {
    if (!this.isTTY) {
      return Promise.reject(new Error("Can't answer a question unless a user TTY"));
    }

    let rl = readline.createInterface({
      input: this.stdin,
      output: this.stdout,
      terminal: true,
    });

    return new Promise((resolve) => {
      this.log(header);

      for (let i = 0; i < options.length; i++) {
        this.log(`${i + 1}. ${options[i]}`);
      }

      let ask = () => {
        rl.question(`${question}?: `, (index) => {
          index = +index;

          if (isNaN(index)) {
            this.log('Not a number');
            ask();
            return;
          }

          if (index <= 0 || index > options.length) {
            this.log('Outside answer range');
            ask();
            return;
          }

          // get index
          index--;
          rl.close();
          resolve(options[index]);
        });
      };

      ask();
    });
  }

  progress(count: number): () => void {
    if (count <= 0) {
      return function() {
        // noop
      };
    }

    if (!this.isTTY) {
      return function() {
        // TODO what should the behaviour here be? we could buffer progress messages maybe
      };
    }

    let bar = new Progress(count, this.stderr);

    bar.render();

    return function() {
      bar.tick();
    };
  }
}
