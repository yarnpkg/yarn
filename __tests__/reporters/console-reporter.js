/* @flow */

import type {MockData} from './_mock.js';
import ProgressBar from '../../src/reporters/console/progress-bar.js';
import Spinner from '../../src/reporters/console/spinner-progress.js';
import ConsoleReporter from '../../src/reporters/console/console-reporter.js';
import build from './_mock.js';

const getConsoleBuff = build(ConsoleReporter, (data): MockData => data);
const stream = require('stream');

// ensures consistency across environments
require('chalk').enabled = true;
require('chalk').supportsColor = true;
require('chalk').styles.blue.open = '\u001b[34m';

test('ConsoleReporter.step', async () => {
  expect(await getConsoleBuff((r) => {
    r.step(1, 5, 'foboar');
  })).toMatchSnapshot();
});

test('ConsoleReporter.header', async () => {
  expect(await getConsoleBuff((r) => {
    r.header('foobar', {name: 'yarn', version: '0.0.0'});
  })).toMatchSnapshot();
});

test('ConsoleReporter.footer', async () => {
  expect(await getConsoleBuff((r) => {
    r.footer(false);
  })).toMatchSnapshot();

  expect(await getConsoleBuff((r) => {
    r.footer(true);
  })).toMatchSnapshot();
});

test('ConsoleReporter.log', async () => {
  expect(await getConsoleBuff((r) => {
    r.log('foobar');
  })).toMatchSnapshot();
});

test('ConsoleReporter.success', async () => {
  expect(await getConsoleBuff((r) => {
    r.success('foobar');
  })).toMatchSnapshot();
});

test('ConsoleReporter.error', async () => {
  expect(await getConsoleBuff((r) => {
    r.error('foobar');
  })).toMatchSnapshot();
});

test('ConsoleReporter.info', async () => {
  expect(await getConsoleBuff((r) => {
    r.info('foobar');
  })).toMatchSnapshot();
});

test('ConsoleReporter.command', async () => {
  expect(await getConsoleBuff((r) => {
    r.command('foobar');
  })).toMatchSnapshot();
});

test('ConsoleReporter.warn', async () => {
  expect(await getConsoleBuff((r) => {
    r.warn('foobar');
  })).toMatchSnapshot();
});

test('ConsoleReporter.tree', async () => {
  const trees = [
    {name: 'dep1'},
    {
      name: 'dep2',
      children: [
        {
          name: 'dep2.1',
          children: [
            {name: 'dep2.1.1'},
            {name: 'dep2.1.2'},
          ],
        },
        {
          name: 'dep2.2',
          children: [
            {name: 'dep2.2.1'},
            {name: 'dep2.2.2'},
          ],
        },
      ],
    },
    {
      name: 'dep3',
      children: [
        {name: 'dep3.1'},
        {name: 'dep3.2'},
      ],
    },
  ];
  expect(await getConsoleBuff((r) => {
    r.tree('', trees);
  })).toMatchSnapshot();
});

test('ConsoleReporter.activity', async () => {
  expect(await getConsoleBuff(function(r) {
    const activity = r.activity();
    activity.tick('foo');
    activity.end();
  })).toMatchSnapshot();
});

test('ConsoleReporter.select', async () => {
  expect(await getConsoleBuff(async function (r, streams): Promise<void> {
    streams.stdin.on('resume', function() {
      streams.stdin.send('1\n', 'ascii');
      streams.stdin.end();
    });

    const res = await r.select('Ayo?', 'Select one', [{
      name: 'foo',
      value: 'foo',
    }, {
      name: 'bar',
      value: 'bar',
    }]);
    expect(res, 'foo');
  })).toMatchSnapshot();
});

test('ConsoleReporter.progress', async () => {
  expect(await getConsoleBuff((r) => {
    r.noProgress = false; // we need this to override is-ci when running tests on ci
    const tick = r.progress(2);
    tick();
    jest.runAllTimers();
    tick();
  })).toMatchSnapshot();

  expect(await getConsoleBuff((r) => {
    const tick = r.progress(0);
    tick();
  })).toMatchSnapshot();

  expect(await getConsoleBuff((r) => {
    r.isTTY = false;
    const tick = r.progress(2);
    tick();
    tick();
  })).toMatchSnapshot();

  expect(await getConsoleBuff((r) => {
    r.noProgress = true;
    const tick = r.progress(2);
    tick();
  })).toMatchSnapshot();
});

test('ProgressBar', () => {
  let data = '';

  class TestStream extends stream.Writable {
    columns: number;
    constructor(options) {
      super(options);
      this.columns = 1000;
    }
    write(chunk: Buffer | string): boolean {
      data += String(chunk);
      return true;
    }
  }
  const bar = new ProgressBar(2, new TestStream());

  bar.render();
  expect(data).toMatchSnapshot();

  bar.tick();
  bar.render();
  expect(data).toMatchSnapshot();

  bar.tick();
  bar.render();
  expect(data).toMatchSnapshot();
});

test('Spinner', () => {
  let data = '';

  class TestStream extends stream.Writable {
    write(chunk: Buffer | string): boolean {
      data += String(chunk);
      return true;
    }
  }
  const spinner = new Spinner(new TestStream());

  spinner.start();
  expect(data).toMatchSnapshot();

  spinner.setText('foo');
  spinner.render();
  expect(data).toMatchSnapshot();

  spinner.setText('bar');
  spinner.render();
  expect(data).toMatchSnapshot();

  spinner.stop();
  expect(data).toMatchSnapshot();
});
