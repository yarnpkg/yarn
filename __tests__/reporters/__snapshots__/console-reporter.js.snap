// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConsoleReporter.activity 1`] = `
Object {
  "stderr": "[1G‚†Å [0K[2K[1G",
  "stdout": "",
}
`;

exports[`ConsoleReporter.command 1`] = `
Object {
  "stderr": "",
  "stdout": "[2K[1G[2m$ foobar[22m",
}
`;

exports[`ConsoleReporter.error 1`] = `
Object {
  "stderr": "[2K[1G[31merror[39m foobar",
  "stdout": "",
}
`;

exports[`ConsoleReporter.footer 1`] = `
Object {
  "stderr": "",
  "stdout": "[2K[1G‚ú®  Done in 0.00s.",
}
`;

exports[`ConsoleReporter.footer 2`] = `
Object {
  "stderr": "",
  "stdout": "[2K[1G‚ú®  Done in 0.00s. Peak memory usage 0.00MB.",
}
`;

exports[`ConsoleReporter.header 1`] = `
Object {
  "stderr": "",
  "stdout": "[2K[1G[1myarn foobar v0.0.0[22m",
}
`;

exports[`ConsoleReporter.info 1`] = `
Object {
  "stderr": "",
  "stdout": "[2K[1G[34minfo[39m foobar",
}
`;

exports[`ConsoleReporter.log 1`] = `
Object {
  "stderr": "",
  "stdout": "[2K[1Gfoobar",
}
`;

exports[`ConsoleReporter.progress 1`] = `
Object {
  "stderr": "[2K[1G[1G‚ñë‚ñë 0/2[1G‚ñà‚ñë 1/2[2K[1G",
  "stdout": "",
}
`;

exports[`ConsoleReporter.progress 2`] = `
Object {
  "stderr": "",
  "stdout": "",
}
`;

exports[`ConsoleReporter.progress 3`] = `
Object {
  "stderr": "",
  "stdout": "",
}
`;

exports[`ConsoleReporter.progress 4`] = `
Object {
  "stderr": "",
  "stdout": "",
}
`;

exports[`ConsoleReporter.select 1`] = `
Object {
  "stderr": "",
  "stdout": "[2K[1G[34minfo[39m Ayo?
[2K[1G  [2m1)[22m foo
[2K[1G  [2m2)[22m bar
[1G[0JSelect one: [13G1",
}
`;

exports[`ConsoleReporter.step 1`] = `
Object {
  "stderr": "",
  "stdout": "[2K[1G[2m[1/5][22m foboar...",
}
`;

exports[`ConsoleReporter.success 1`] = `
Object {
  "stderr": "",
  "stdout": "[2K[1G[32msuccess[39m foobar",
}
`;

exports[`ConsoleReporter.warn 1`] = `
Object {
  "stderr": "[2K[1G[33mwarning[39m foobar",
  "stdout": "",
}
`;

exports[`ProgressBar 1`] = `"[2K[1G[1G‚ñë‚ñë 0/2"`;

exports[`ProgressBar 2`] = `"[2K[1G[1G‚ñë‚ñë 0/2[1G‚ñà‚ñë 1/2"`;

exports[`ProgressBar 3`] = `"[2K[1G[1G‚ñë‚ñë 0/2[1G‚ñà‚ñë 1/2[2K[1G[1G‚ñà‚ñà 2/2"`;

exports[`Spinner 1`] = `"[1G‚†Å [0K"`;

exports[`Spinner 2`] = `"[1G‚†Å [0K[1G‚†Ç foo[0K"`;

exports[`Spinner 3`] = `"[1G‚†Å [0K[1G‚†Ç foo[0K[1G‚†Ñ bar[0K"`;

exports[`Spinner 4`] = `"[1G‚†Å [0K[1G‚†Ç foo[0K[1G‚†Ñ bar[0K[2K[1G"`;
